src/backend/services/chat.py

```diff
@@ -174,6 +158,65 @@ def process_chat(
     )
 
 
+def check_chat_request_tools(
+    chat_request: BaseChatRequest, entity: Agent | User
+) -> None:
+    if chat_request.tools:
+        entity_type = "agent" if isinstance(entity, Agent) else "user"
+        for tool in chat_request.tools:
+            if tool.name not in entity.tools:
+                raise HTTPException(
+                    status_code=404,
+                    detail=f"Tool {tool.name} not found in {entity_type} {entity.id}",
+                )
+
+
+def process_user_agents_tools(
+    session: DBSessionDep, chat_request: BaseChatRequest, ctx: Context
+) -> None:
+    """
+    Process user's, agent's tools for a chat request.
+
+    Args:
+        chat_request (BaseChatRequest): Chat request data.
+        ctx (Context): Context object.
+
+    Returns:
+        dict: Dictionary containing the processed tools.
+    """
+    user_id = ctx.get_user_id()
+    agent_id = ctx.get_agent_id()
+    user = None
+    agent = None
```

Comment:
We probably don't need to call `get_user_id()` and `get_agent_id()`, you can shorten it to 
```
ctx.with_user(session)
user = ctx.get_user()
```
and directly check if `user is None`

Reply:
Same thing for agent

Reply:
I need these IDs for the exception details